test/Npgsql.Tests/AuthenticationTests.cs

```diff
@@ -0,0 +1,441 @@
+
+namespace Npgsql.Tests;
+
+public class AuthenticationTests : MultiplexingTestBase
+{
+    [Test]
+    [NonParallelizable] // Sets environment variable
+    public async Task Connect_UserNameFromEnvironment_Succeeds()
+    {
+        var builder = new NpgsqlConnectionStringBuilder(ConnectionString) { IntegratedSecurity = false };
+        using var _ = SetEnvironmentVariable("PGUSER", builder.Username);
+        builder.Username = null;
+        using var __ = CreateTempPool(builder.ConnectionString, out var connectionString);
+        using var ___ = await OpenConnectionAsync(connectionString);
+    }
+
+    [Test]
+    [NonParallelizable] // Sets environment variable
+    public async Task Connect_PasswordFromEnvironment_Succeeds()
+    {
+        var builder = new NpgsqlConnectionStringBuilder(ConnectionString) { IntegratedSecurity = false };
+        using var _ = SetEnvironmentVariable("PGPASSWORD", builder.Password);
+        builder.Password = null;
+        using var __ = CreateTempPool(builder.ConnectionString, out var connectionString);
+        using var ___ = await OpenConnectionAsync(connectionString);
+    }
+
+    [Test]
+    public async Task Set_Password_on_NpgsqlDataSource()
+    {
+        var dataSourceBuilder = GetPasswordlessDataSourceBuilder();
+        await using var dataSource = dataSourceBuilder.Build();
+
+        // No password provided
+        Assert.That(() => dataSource.OpenConnectionAsync(), Throws.Exception.TypeOf<NpgsqlException>());
+
+        var connectionStringBuilder = new NpgsqlConnectionStringBuilder(TestUtil.ConnectionString);
+        dataSource.Password = connectionStringBuilder.Password!;
+
+        await using var connection1 = await dataSource.OpenConnectionAsync();
+        await using var connection2 = dataSource.OpenConnection();
+    }
+
+    [Test]
+    public async Task Periodic_password_provider()
+    {
+        var dataSourceBuilder = GetPasswordlessDataSourceBuilder();
+        var password = new NpgsqlConnectionStringBuilder(TestUtil.ConnectionString).Password!;
+
+        var mre = new ManualResetEvent(false);
+        dataSourceBuilder.UsePeriodicPasswordProvider((_, _) =>
+        {
+            mre.Set();
+            return new(password);
+        }, TimeSpan.FromMilliseconds(100), TimeSpan.FromSeconds(10));
+
+        await using (var dataSource = dataSourceBuilder.Build())
+        {
+            await using var connection1 = await dataSource.OpenConnectionAsync();
+            await using var connection2 = dataSource.OpenConnection();
+
+            mre.Reset();
+            if (!mre.WaitOne(TimeSpan.FromSeconds(30)))
+                Assert.Fail("Periodic password refresh did not occur");
+        }
+
+        mre.Reset();
+        if (mre.WaitOne(TimeSpan.FromSeconds(1)))
+            Assert.Fail("Periodic password refresh occurred after disposal of the data source");
+    }
+
+    [Test]
+    public async Task Periodic_password_provider_with_first_time_exception()
```

Comment:
Potentially useful to add a test with a provider failing only the second time as well (for example with an infinite success interval and a 1 second failure interval)

Reply:
The problem is that with an infinite success interval, there's never a second time for the failure :)

But yeah, a test that we continue with the last known password after an exception makes sense, will do.